### PR TITLE
[feature] #2407: Parametrise triggers

### LINKED ISSUE #2407: Parametric smartcontracts
It should be possible to forward the event that triggered the execution of a trigger into the body of the trigger, at least in the WASM format.

### PR DESCRIPTION
**Pull request by Arjentix**
### Description of the Change

* Added event storing in `trigger::Set`
* `WorldStateView` now passes event from `trigger::Set` to `wasm::Runtime`
* `wasm::Runtime` was a little bit refactored and got a new function `execute_trigger()`
* Added `trigger_entrypoint` macro to `iroha_wasm` which accepts function with `event` parameter
* `mint_nft_for_every_user_smartcontract` was updated to use `trigger_entrypoint`

### Issue

* Closes #2407
* After merging this PR we should discuss possibility to make event strongly typed

### Benefits

* Now smartcontract writers can handle an event, that triggered thier smartcontract execution

### Possible Drawbacks

Events are not strong-typed. A top-level enum `Event` is used for now, which is not very handy.
I have some ideas, how to fix ths, but it needs further disscussion

### Usage Examples or Tests 

1. Comment line **173** (`#[ignore = "Only on nightly"]`) in `client/tests/integration/triggers/time_triggers.rs`
2. Run
    ```bash
    cargo +nightly-2022-04-20 test --package iroha_client --test mod --integration::triggers::time_trigger::mint_nft_for_every_user_every_1_sec --exact --nocapture
    ```

### CODE REVIEW

**Comment by Arjentix** [comment]
@appetrosyan , @mversic , I'm hopping to discuss with you event strong-typing later. I have some ideas

File: data_model/src/trigger/set.rs
Comment (outoftardis) [inline data_model/src/trigger/set.rs:340]:
```suggestion
    /// Match and insert a [`TriggerId`] into the set of matched ids.
```

File: wasm/derive/src/lib.rs
Comment (outoftardis) [inline wasm/derive/src/lib.rs:13]:
```suggestion
/// Used to annotate the user-defined function that starts the execution of a smart contract.
```

File: wasm/derive/src/lib.rs
Comment (outoftardis) [inline wasm/derive/src/lib.rs:28]:
```suggestion
/// Used to annotate the user-defined function that starts the execution of a trigger.
///
/// Should be used for trigger smart contracts.
/// For just smart contract (for transactions), see [`entrypoint`].
```

File: wasm/derive/src/lib.rs
Comment (outoftardis) [inline wasm/derive/src/lib.rs:16]:
```suggestion
/// Should be used for smart contracts (inside transactions).
/// For triggers, see [`trigger_entrypoint`].
```

File: core/src/smartcontracts/wasm.rs
Comment (outoftardis) [inline core/src/smartcontracts/wasm.rs:582]:
```suggestion
    /// Return the offset and length of the encoded object
```

File: core/src/smartcontracts/wasm.rs
Comment (outoftardis) [inline core/src/smartcontracts/wasm.rs:481]: will you add the description before merging?

File: core/src/smartcontracts/wasm.rs
Comment (outoftardis) [inline core/src/smartcontracts/wasm.rs:46]:
```suggestion
    /// Name of the imported function to debug print objects
```

File: core/src/smartcontracts/wasm.rs
Comment (outoftardis) [inline core/src/smartcontracts/wasm.rs:35]:
```suggestion
    /// Name of the exported entry for trigger execution
```

File: core/src/smartcontracts/wasm.rs
Comment (outoftardis) [inline core/src/smartcontracts/wasm.rs:33]:
```suggestion
    /// Name of the exported entry for smartcontract (not trigger) execution
```

File: core/src/smartcontracts/wasm.rs
Comment (Arjentix) [reply core/src/smartcontracts/wasm.rs:481]: Oh, completely forgot about this, thanks

File: wasm/derive/src/lib.rs
Comment (mversic) [inline wasm/derive/src/lib.rs:18]: I think it's good as is. I would remove the comment

File: client/tests/integration/create_nft_for_every_user_smartcontract/src/lib.rs
Comment (mversic) [inline client/tests/integration/create_nft_for_every_user_smartcontract/src/lib.rs:18]: another possibility would have been to let smartcontract ask wsv about the context of it's execution. Here this would mean to query which event triggered it's execution. The same could be done with `AccountId`. That's pull vs push architecture. I'm not strongly advocating one vs the other, I'm just thinking what would be easier for the writer to comprehend. Especially, if the number of args in the input will grow in the future and if they'll be rarely used

File: core/src/smartcontracts/wasm.rs
Comment (mversic) [inline core/src/smartcontracts/wasm.rs:39]: I would name it `import` and `export`

File: core/src/smartcontracts/wasm.rs
Comment (mversic) [inline core/src/smartcontracts/wasm.rs:303]:
```suggestion
        let query = Self::decode_from_memory(&mut caller, offset, len)?;
```

seems to me that you could do it this way

File: core/src/smartcontracts/wasm.rs
Comment (mversic) [inline core/src/smartcontracts/wasm.rs:589]:
```suggestion
    fn encode_into_memory<T: Encode>(
```

File: core/src/smartcontracts/wasm.rs
Comment (mversic) [reply core/src/smartcontracts/wasm.rs:303]: also, can you use `encode_to_memory` when writing into memory inside this function?

File: core/src/smartcontracts/wasm.rs
Comment (mversic) [inline core/src/smartcontracts/wasm.rs:560]:
```suggestion
            Self::encode_to_memory(account_id, &alloc_fn, &mut store)?
```
also seems to me that you can get memory from inside the function

Comment (mversic) [reply core/src/smartcontracts/wasm.rs:560]: although it's a little bit weird that you get memory from caller when decoding and from store when encoding. As if something's not quite right with abstraction. Idk, if it can be made better

**Review by mversic — CHANGES_REQUESTED** [review]

File: client/tests/integration/create_nft_for_every_user_smartcontract/src/lib.rs
Comment (appetrosyan) [reply client/tests/integration/create_nft_for_every_user_smartcontract/src/lib.rs:18]: Personally I think that we can do a bit of both. The number of parameters of this function can be different, and we can use the macros to generate the query of the WSV as if it were a parameter, and only add the event to the smartcontract if we need it (which we can tell from the macro by either not having the parameter at all, or having it unused).

File: core/src/smartcontracts/wasm.rs
Comment (appetrosyan) [reply core/src/smartcontracts/wasm.rs:560]: This is a code smell, but I doubt that it can leak memory. 

New ticket?

File: client/tests/integration/create_nft_for_every_user_smartcontract/src/lib.rs
Comment (appetrosyan) [reply client/tests/integration/create_nft_for_every_user_smartcontract/src/lib.rs:18]: As a bonus you now have only one `entrypoint` macro

File: data_model/src/trigger/set.rs
Comment (appetrosyan) [inline data_model/src/trigger/set.rs:280]: This comment just repeats the if. Could just remove it.

File: wasm/derive/src/lib.rs
Comment (appetrosyan) [inline wasm/derive/src/lib.rs:25]:
```suggestion
/// Use to annotate the user-defined function that starts the execution of a trigger.
```

File: wasm/derive/src/lib.rs
Comment (appetrosyan) [reply wasm/derive/src/lib.rs:49]: Add a Ui test. See if this check can be simplified (removed).

Comment (appetrosyan) [reply wasm/derive/src/lib.rs:49]: Also, we're mixing levels of abstraction here. If we allow a path as a type specification, we need to make sure that the entire path agrees, not just the last segment. That way you can't confuse `my::AccountId` with the expected type. 

Doing this properly is a lot of work. 

Instead consider offloading the work onto the built-in type checker. Remove this check and ensure that the type is enforced in the generated code.

File: wasm/derive/src/smartcontract.rs
Comment (appetrosyan) [inline wasm/derive/src/smartcontract.rs:39]: This doesn't look right.

File: wasm/derive/src/lib.rs
Comment (Arjentix) [reply wasm/derive/src/lib.rs:49]: @mversic , you are right, it's just a peace of old code I forgot to update

Comment (Arjentix) [reply wasm/derive/src/lib.rs:49]: > Instead consider offloading the work onto the built-in type checker. Remove this check and ensure that the type is enforced in the generated code.

Good point, I'll do that

Comment (Arjentix) [reply wasm/derive/src/lib.rs:49]: Although this will make an error message less descriptive

File: client/tests/integration/create_nft_for_every_user_smartcontract/src/lib.rs
Comment (Arjentix) [reply client/tests/integration/create_nft_for_every_user_smartcontract/src/lib.rs:18]: So do you suggest to make entrypoint function parameter-less? And query all needed parameters from fn body? To me it sounds like a step away from strong-typing from user point of view. I think an *event* which have triggered smartcontract execution will be usefull pretty always

File: core/src/smartcontracts/wasm.rs
Comment (Arjentix) [reply core/src/smartcontracts/wasm.rs:303]: > seems to me that you could do it this way

To retrieve `memory` from `caller` we need to call `Self::get_memory()` which will try to find the exported memory. If I'm gonna to do that in `decode_from_memory()` I have to call `get_memory()`. And if some function wants to decode two objects from the same `memory` then there will be **2** calls for memory exporting instead of one. Exporting looks like a big job, so I decided to add a `memory` parameter.

File: core/src/smartcontracts/wasm.rs
Comment (Arjentix) [reply core/src/smartcontracts/wasm.rs:560]: > also seems to me that you can get memory from inside the function

See [this comment](https://github.com/hyperledger/iroha/pull/2569#discussion_r936034813).
Also as an example of writing two times see `execute_trigger()`

Comment (Arjentix) [reply core/src/smartcontracts/wasm.rs:560]: > although it's a little bit weird that you get memory from caller when decoding and from store when encoding. As if something's not quite right with abstraction. Idk, if it can be made better

Hm... All I've done is just refactoring by looking and the previous existed code and trying to avoid repeating it by encapsulating in functions when I was writing the `execute_trigger()` function. I'm not really sure what does `store` means here. If you have some ideas, as Aleksandr suggested, let's create a new issue. I'm just have no idea how to describe this issue

File: wasm/derive/src/smartcontract.rs
Comment (Arjentix) [reply wasm/derive/src/smartcontract.rs:39]: Fixed. I just moved that from the old code

Comment (Arjentix) [reply wasm/derive/src/smartcontract.rs]: From the smartcontract writer point of view it makes sense to pass parameters by values. It may use them or may not

File: core/src/smartcontracts/wasm.rs
Comment (mversic) [reply core/src/smartcontracts/wasm.rs]: I think it's ok

File: client/tests/integration/create_nft_for_every_user_smartcontract/src/lib.rs
Comment (appetrosyan) [reply client/tests/integration/create_nft_for_every_user_smartcontract/src/lib.rs:18]: This is relatively similar to how Rust handles CLI arguments. Most other languages accept `int argc, char** argv`, but Rust has a parameter-less entrypoint, so what we're doing wouldn't be surprising to seasoned Rust engineers.

Comment (appetrosyan) [reply client/tests/integration/create_nft_for_every_user_smartcontract/src/lib.rs:18]: But we can also make a concession to `Event` as an argument. 

Most executables are parameter-less, so for them the `iroha_wasm::entrypoint` behaviour doesn't change. 

Some of the trigger executables need the `Event`, but most don't, which is why we didn't have it as part of the original implementation. So `iroha_wasm::trigger_entrypoint` could parse the extra argument, and simply be syntactic sugar around the query in the `fn` body. That way, if you don't need an event, you don't waste time moving it from the outers scope.

File: core/src/smartcontracts/wasm.rs
Comment (appetrosyan) [reply core/src/smartcontracts/wasm.rs]: I think this will go away if instead of pushing the event into triggers unconditionally, we instead wrap syntactic sugar around querying.

File: client/tests/integration/create_nft_for_every_user_smartcontract/src/lib.rs
Comment (mversic) [reply client/tests/integration/create_nft_for_every_user_smartcontract/src/lib.rs:18]: > To me it sounds like a step away from strong-typing from user point of view.

I don' think this would affect typing

> Fortunately, the number of parameters of the entrypoint function can vary without changing the number of parameters in the generated function.

I would take this suggestion which leaves the smartcontract API pretty much unaffected. In this way user would never manually execute queries like `get_account_id` or `get_event`, instead the macro would guess the queries from the function signature. Also, there should be no need for `needless_pass_by_value` anymore

**Comment by Arjentix** [comment]
After discussion we decided to implement *pull* architecture. I'm converting this PR into draft for now

**Comment by Arjentix** [comment]
While updating this PR I also have tried to implement some tests for my procedural macro using `trybuild` crate.
I've spent a lot of time and realised that I simply can't do that because of problems with target architecture desired by `iroha_wasm` and problems with `std` and `no_std` depenencies.

I think, someday we will need tests for `iroha_wasm` and `iroha_wasm_derive` crates, because they are an important part of our API for external users. We need to do a research and implement it somehow.

Interesting note: `trybuild` does not activate `cfg(test)` for your macro

**Review by mversic — CHANGES_REQUESTED** [review]

File: wasm/derive/src/lib.rs
Comment (mversic) [inline wasm/derive/src/lib.rs:36]: why the underscores in front?

File: wasm/derive/src/lib.rs
Comment (mversic) [inline wasm/derive/src/lib.rs:223]: would it be possible to do the conversion before calling the smart contract, not inside of it. I wouldn't mind too much, had it not made you mess up the dbg macro. Also, it's better to move out of smart contract as much code as possible

File: core/src/smartcontracts/wasm.rs
Comment (mversic) [inline core/src/smartcontracts/wasm.rs:507]:
```suggestion
        self.execute_with_state(bytes, state)
```

maybe this?

Comment (mversic) [inline core/src/smartcontracts/wasm.rs:507]:
```suggestion
        self.execute_with_state(bytes, state)
```

File: wasm/derive/src/lib.rs
Comment (mversic) [reply wasm/derive/src/lib.rs:223]: what you're saying is that executing a smart contract is valid only for a certain type of event. If you registered this smart contract for some other event the execution would fail. 

My questions are:
1. Why is it not ok for a smart contract to be able to be triggered on multiple events?
2. Is this covered in the validation step? I wouldn't like to see that we've successfully registered a trigger but when the actual event occurs the execution of smart contract fails (I'm not sure how trigger validation works).

File: wasm/derive/src/lib.rs
Comment (Arjentix) [reply wasm/derive/src/lib.rs:36]: Because this field is not used anywhere and clippy complains about it. As far as I remember it's a common practice for structures like that (i.e. `actor/derive/` and `config/derive/`)

File: wasm/derive/src/lib.rs
Comment (Arjentix) [reply wasm/derive/src/lib.rs:223]: > Is this covered in the validation step? I wouldn't like to see that we've successfully registered a trigger but when the actual event occurs the execution of smart contract fails (I'm not sure how trigger validation works).

No yet covered. But yes, I agree, I'd like to get the error about incompatible event type at `Register` step. This is a subject for the next PR, as we discussed with @appetrosyan.

> Why is it not ok for a smart contract to be able to be triggered on multiple events?

There are two notes:
1. Currently we provide `Filter` at trigger registration step. `Filter` can work only with one concrete type of events. That's why our trigger can't receive different event types for now. Maybe this will be changed once we have better filters. That's probably a question for @appetrosyan .
2. You still can use top-level event, because there is always `impl TryInto<T> for T`.

File: wasm/src/debug.rs
Comment (Arjentix) [reply wasm/src/debug.rs:114]: Hm..., I'm looking at it more like [`colored crate`](https://crates.io/crates/colored) rather than like `proc_macro_error`.

With `colored` you can write code like `println!("foo".blue());`. This will display `foo` using blue color. But if you enable feature `no-color`, then this `blue()` will do nothing. This is usefull, because you can have beautiful output for normal terminals and have optional feature `dumb_terminal = ["colored/no-color"]` for terminals without color support.

The same thing here. I'd like the users to keep their code the same and only manipulate with feature flag to enable/disable debugging.

`expect_or_abort` has different meaning -- it will always abort on `Err/None` variant.

File: core/src/smartcontracts/wasm.rs
Comment (Arjentix) [reply core/src/smartcontracts/wasm.rs:564]: Good idea

File: core/src/smartcontracts/wasm.rs
Comment (Arjentix) [reply core/src/smartcontracts/wasm.rs:32]: Indeed, that's a bug. I'll use this one and remove `_iroha_trigger_main`

File: core/src/smartcontracts/wasm.rs
Comment (Arjentix) [reply core/src/smartcontracts/wasm.rs:507]: You are right, didn't mentioned that before

File: wasm/derive/src/lib.rs
Comment (mversic) [reply wasm/derive/src/lib.rs:223]: Alternative would be to support only `Event` as `entrypoint` argument and leave it to the writer of smart contract to make the conversion. Not sure if I would prefer it, but it kinda makes sense because you are essentially faking strong typing in the `entrypoint` by implicitly doing the conversion

Comment (Arjentix) [reply wasm/derive/src/lib.rs:223]: Well, I think, all strong-typing with WASM here actually would be faking. You can always send something different instead of i.e. `AccountId` and you won't notice that until runtime.
Here is the same thing. How can this typing be stronger (especially with *pull*-architecture).
In terms of implicit conversion. I think, your approach will allow only one way -- use top-level event. And with my approach user can use any compatible event, including top-level

Comment (Arjentix) [reply wasm/derive/src/lib.rs:223]: Let's hear more voices here. Democracy!

File: wasm/src/debug.rs
Comment (Arjentix) [reply wasm/src/debug.rs:114]: Let's summon more voices here as well

**Review by mversic — DISMISSED** [review]

File: wasm/derive/src/lib.rs
Comment (appetrosyan) [reply wasm/derive/src/lib.rs:223]: Guys, the typing arguments are all well and good, but we should agree on a few things:

1. We need this soon. ASAP. 
2. If we enforce top-level events first and then keeping the API the same, change the underlying behaviour, we can have a smooth transition. 
3. Allowing anything but the top-level events is UB. We can define that behaviour later.

Comment (Arjentix) [reply wasm/derive/src/lib.rs:223]: I don't fully understand, what do you mean, but okay. If we need this ASAP, when we need just one more approval here

**Review by appetrosyan — DISMISSED** [review]

**Review by mversic — APPROVED** [review]

**Review by appetrosyan — APPROVED** [review]

**Pull request merged**
